lib/r10k/tarball.rb

```diff
@@ -0,0 +1,257 @@
+require 'digest'
+require 'fileutils'
+require 'find'
+require 'minitar'
+require 'net/http'
+require 'tempfile'
+require 'uri'
+require 'zlib'
+require 'r10k/settings'
+require 'r10k/settings/mixin'
+require 'r10k/util/platform'
+
+module R10K
+  class Tarball
+
+    include R10K::Settings::Mixin
+
+    def_setting_attr :proxy
+    def_setting_attr :cache_root, if R10K::Util::Platform.windows?
+                                    File.join(ENV['LOCALAPPDATA'], 'r10k', 'tarball')
+                                  else
+                                    File.expand_path(ENV['HOME'] ? '~/.r10k/tarball': '/root/.r10k/tarball')
+                                  end
+
+    # @!attribute [r] name
+    #   @return [String] The tarball's name
+    attr_accessor :name
+
+    # @!attribute [r] source
+    #   @return [String] The tarball's source
+    attr_accessor :source
+
+    # @!attribute [r] sha256digest
+    #   @return [String] The tarball's expected sha256 digest
+    attr_accessor :sha256digest
```

Comment:
Honestly, I'm leaning towards `version` being the term that gets documented.

If we shorten it, I think `digest` (or the commonly used but slightly more technically ambiguous `checksum`) would be the preferred shortening. That's the part that conveys what this parameter is really about—the output of a hash function which can be used to validate the expected content.

I went with the very specific and technical `sha256digest` because I was trying to avoid the user needing to look up which algorithm to use for the `checksum`. However, we could probably expose that later by offering a `checksum_algorithm` if a use case arose for it.

I've put in a commit to change this to just `checksum` for now.